images/encryption/blockcipher/blockcipher.go

```diff
@@ -0,0 +1,111 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package blockcipher
+
+import (
+	"io"
+
+	"github.com/containerd/containerd/content"
+	"github.com/pkg/errors"
+)
+
+// CryptedDataReader wraps the io.Reader and adds the Size of the encrypted or decrypted
+// data to it as well as its hash
+type CryptedDataReader interface {
+	content.ReaderDigester
+}
+
+// LayerCipherType is the ciphertype as specified in the layer metadata
+type LayerCipherType string
+
+// TODO: Should be obtained from OCI spec once included
+const (
+	AESSIVCMAC256 LayerCipherType = "AES-SIV-CMAC-256"
+	AESSIVCMAC512 LayerCipherType = "AES-SIV-CMAC-512"
+	CipherTypeOpt string          = "type"
+)
+
+// LayerBlockCipherOptions includes the information required to encrypt/decrypt
+// an image
+type LayerBlockCipherOptions struct {
+	SymmetricKey  []byte            `json:'symkey'`
+	CipherOptions map[string][]byte `json:'cipheroptions'`
+}
+
+// LayerBlockCipher returns a provider for encrypt/decrypt functionality
+// for handling the layer data for a specific algorithm
+type LayerBlockCipher interface {
+	// GenerateKey creates a symmetric key
+	GenerateKey() []byte
+	// Encrypt takes in layer data and returns the ciphertext and relevant LayerBlockCipherOptions
+	Encrypt(layerDataReader io.ReaderAt, opt LayerBlockCipherOptions) (CryptedDataReader, LayerBlockCipherOptions, error)
+	// Decrypt takes in layer ciphertext data and returns the plaintext and relevant LayerBlockCipherOptions
+	Decrypt(layerDataReader io.ReaderAt, opt LayerBlockCipherOptions) (CryptedDataReader, LayerBlockCipherOptions, error)
+}
+
+// LayerBlockCipherHandler is the handler for encrypt/decrypt for layers
+type LayerBlockCipherHandler struct {
+	cipherMap map[LayerCipherType]LayerBlockCipher
+}
+
+// Encrypt is the handler for the layer decryption routine
+func (h *LayerBlockCipherHandler) Encrypt(plainDataReader io.ReaderAt, typ LayerCipherType) (CryptedDataReader, LayerBlockCipherOptions, error) {
+
+	if c, ok := h.cipherMap[typ]; ok {
+		opt := LayerBlockCipherOptions{
+			SymmetricKey: c.GenerateKey(),
+		}
+		encDataReader, newopt, err := c.Encrypt(plainDataReader, opt)
+		if err == nil {
+			newopt.CipherOptions[CipherTypeOpt] = []byte(typ)
+		}
+		return encDataReader, newopt, err
+	}
+	return nil, LayerBlockCipherOptions{}, errors.New("Unsupported cipher type")
```

Comment:
nit: error string starts with a lower character

Reply:
I pushed a patch that now fixes this throughout the PR: https://github.com/containerd/containerd/pull/3134/commits/f9e9df18b89589cf225c99955edf63022e3c1588